### PR TITLE
Split CI workflow

### PR DESCRIPTION
**Pull request by michaljurecko**
Part of: https://keboola.atlassian.net/browse/BA-33

Changes:
- One huge workflow split to more reusable workflows.
- Removed `cross-compile` step, compilation to Linux/MacOS/Windows is tested in CLI E2E tests.
- Added projects with `"stagingStorage":"abs"`.
- 20-40% speedup.
- On CLI/API release, tests for E2E API/CLI don't have to run.
- Prepared empty workflow for the Buffer service release.

------
:cry: unfortunately, I did not manage to solve the `concurrency`, so GHA still kills older pending jobs.

------
image

image

### CODE REVIEW

File: .github/actions/go-cache/action.yml [hunk 1]
```diff
@@ -17,15 +17,14 @@ runs:
...
         path: |
-          ~/go/pkg/mod
-          ~\go\pkg\mod
-          ~/.cache/go-build
-          ~/Library/Caches/go-build
-          ~\AppData\Local\go-build
+          ~/go/pkg
+          ~\go\pkg
+          ~/go/bin
+          ~\go\bin
```
Comment (michaljurecko) [inline .github/actions/go-cache/action.yml:23]: Cache ma `100M` namiesto `1GB`, hlavne na windows je stahovanie cache velmi pomale, tak dava radsej zmysel cachovat iba dependencies, a nie casti kompilacie.

File: .github/actions/push-image-aws-azure/action.yml [hunk 2]
```diff
@@ -0,0 +1,71 @@
+name: 'Push image to Amazon ECR and Azure ACR'
+description: 'Push image to Amazon ECR and Azure ACR'
```
Comment (michaljurecko) [inline .github/actions/push-image-aws-azure/action.yml:2]: Spravil som samostatnu action z uploadu image do ECR, ACR.

V Buffer service bude image pre API a Worker + image pre Templates API:
Tj. bude to pouzite na 3 miestach.

Original:
https://github.com/keboola/keboola-as-code/blob/fe5f1bff3bec0e5bb3251394e9e25925e5f1122b/.github/workflows/push.yml#L634-L687

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,30 @@
+name: 'Push CI'
+
+on:
+  push:
+
+jobs:
+  test-lint:
+    name: "Lint"
+    secrets: inherit
+    uses: ./.github/workflows/test-lint.yml
```
Comment (michaljurecko) [inline .github/workflows/ci.yml:10]: Takto pekne si to vieme skladat. Tu je CI co sa spusta pri kazdom pushe.

Vo workflow pre release napr. nemusime spustat API E2E testy, ked releasujeme CLI.

File: .github/workflows/provisioning-cli-s3.yml [hunk 4]
```diff
@@ -1,5 +1,7 @@
-name: 'CLI Dist Provisioning'
+name: 'Provisioning: CLI S3 Distribution'
+
 on:
+  workflow_call:
```
Comment (michaljurecko) [inline .github/workflows/provisioning-cli-s3.yml:4]: Okrem samostatneho volania pri zmene, to vieme zavolat aj vramci CLI release.

File: .github/workflows/release-cli.yml [hunk 5]
```diff
@@ -600,12 +420,9 @@ jobs:
         run: |
           brew install keboola/keboola-cli/keboola-cli
           kbc --version | tee /dev/stderr | grep -q "Version:    ${{ env.VERSION }}"
-        env:
-          VERSION: ${{ needs.set-version.outputs.version }}
```
Comment (michaljurecko) [inline .github/workflows/release-cli.yml:604]: VERSION a TAG su globalne env hore.

File: .github/workflows/release-service-buffer.yml [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    permissions:
+      id-token: write
+    needs:
+      - test-lint
+      - test-unit
+      - test-e2e-service-templates
+      - test-k8s-service-templates
+    steps:
+      - name: TODO
+        run: echo "TODO"
```
Comment (michaljurecko) [inline .github/workflows/release-service-buffer.yml:48]: Priprava pre buffer API.

File: .github/workflows/release-service-templates.yml [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        with:
+          imageName: ${{ env.IMAGE_NAME }}
+          imageTag: ${{ env.IMAGE_TAG }}
+          ecrRegion: ${{ env.ECR_REGION }}
+          ecrRepository: ${{ env.ECR_REPOSITORY }}
+          ecrPushRole: ${{ env.ECR_PUSH_ROLE }}
+          acrRepository: ${{ env.ACR_REPOSITORY }}
+          acrRegistry: ${{ env.ACR_REGISTRY }}
+          acrUsername: ${{ env.ACR_USERNAME }}
+          acrPassword: ${{ secrets.ACR_PASSWORD }}
```
Comment (michaljurecko) [inline .github/workflows/release-service-templates.yml:73]: Pouzitie novej action.

File: Makefile [hunk 8]
```diff
@@ -50,25 +50,25 @@ tests-verbose: prepare
...
 tests-templates-api: generate-templates-api
-	TEST_VERBOSE=false TEST_LOG_FORMAT=standard-verbose TEST_PACKAGE=./test/api/templates/... bash ./scripts/tests.sh -run TestTemplatesApiE2E
+	TEST_PACKAGE=./test/api/templates/... bash ./scripts/tests.sh -run TestTemplatesApiE2E
 
 tests-templates-api-verbose: generate-templates-api
 	TEST_VERBOSE=true TEST_LOG_FORMAT=standard-verbose TEST_PARALLELISM=1 TEST_PARALLELISM_PKG=1 TEST_PACKAGE=./test/api/templates/... bash ./scripts/tests.sh -run TestTemplatesApiE2E
 
 tests-buffer-api: generate-buffer-api
-	TEST_VERBOSE=false TEST_LOG_FORMAT=standard-verbose TEST_PACKAGE=./test/api/buffer/... bash ./scripts/tests.sh -run TestBufferApiE2E
+	TEST_PACKAGE=./test/api/buffer/... bash ./scripts/tests.sh -run TestBufferApiE2E
```
Comment (michaljurecko) [inline Makefile:71]: Tieto ciastkove testy volame teraz v CI, mali nastaveny iny format vystupu.

File: docker-compose.yml [hunk 9]
```diff
@@ -11,12 +11,8 @@ services:
       - ./:/code:z
       - cache:/tmp/cache
     environment:
-      # For CLI
+      # For all
       - TEST_KBC_PROJECTS
-      ## OR
-      - TEST_KBC_PROJECT_ID
-      - TEST_KBC_STORAGE_API_HOST
-      - TEST_KBC_STORAGE_API_TOKEN
```
Comment (michaljurecko) [inline docker-compose.yml:19]: Uz to nepodporujeme.

File: internal/pkg/api/server/templates/dependencies/locks_test.go [hunk 10]
```diff
@@ -148,19 +148,19 @@ func newTestEtcdClient(t *testing.T) *etcd.Client {
 	assert.NoError(t, err)
 
 	// Check if etcd is enabled
-	if envs.Get("ETCD_ENABLED") == "false" {
+	if envs.Get("TEMPLATES_API_ETCD_ENABLED") == "false" {
```
Comment (michaljurecko) [inline internal/pkg/api/server/templates/dependencies/locks_test.go:151]: Tu este neboli upravene nazvy ENVs, aj v CI boli stare.

File: scripts/tests.sh [hunk 11]
```diff
@@ -13,7 +13,7 @@ pwd
 
 TEST_VERBOSE="${TEST_VERBOSE:=false}"
 TEST_PARALLELISM="${TEST_PARALLELISM:=8}"      # number of parallel tests inside package
-TEST_PARALLELISM_PKG="${TEST_PARALLELISM:=8}"  # number of packages tested in parallel
+TEST_PARALLELISM_PKG="${TEST_PARALLELISM_PKG:=8}"  # number of packages tested in parallel
```
Comment (michaljurecko) [inline scripts/tests.sh:16]: Typo

File: scripts/tools.sh [hunk 12]
```diff
@@ -9,9 +9,38 @@ set -o pipefail         # Use last non-zero exit code in a pipeline
...
 cd "$SCRIPT_DIR"
 
-./install-gotestsum.sh -b $(go env GOPATH)/bin
-./install-goreleaser.sh -b $(go env GOPATH)/bin v0.182.1
-curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(go env GOPATH)/bin v1.50.1
-curl -sSfL https://raw.githubusercontent.com/cosmtrek/air/master/install.sh | sh -s -- -b $(go env GOPATH)/bin
-go install golang.org/x/tools/cmd/godoc@latest
-go install goa.design/goa/v3/cmd/goa@v3
+# gotestsum
+if ! command -v gotestsum &> /dev/null
```
Comment (michaljurecko) [inline scripts/tools.sh:13]: V CI pouzivame cache, teda nemusi instalovat `tools` znova, ak binarky uz existuju.

Inak, idealne by bolo pouzit vsade `dev` image, ale WIndows a MacOs v GHA nepodporuju nativne images, zatial.

**Comment by michaljurecko** [comment]
Note: urcite to niekde v release Templates API alebo CLI spadne (ako mame skusenost so zmenami v pipeline), ale to uvidime az ked sa to vyskusa.

**Review by JakubMatejka — APPROVED** [review]
Super👍 Už jsem si taky delší dobu říkal že by to chtělo. 🙂

**Pull request merged**